docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+    ImmutableArray<SymbolDisplayPart> ToDisplayParts(Nullability nullability, SymbolDisplayFormat format = null);
+    string ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+    ImmutableArray<SymbolDisplayPart> ToMinimalDisplayString(SemanticModel semanticModel, int position, Nullability nullability, SymbolDisplayFormat format = null);
+}
+
+/*
+ * There will be more API changes, such as on SyntaxFactory and SyntaxGenerator. This section will be updated later when we have designs for this ready.
+ */
+
+#endregion
+```
+
+## Considered Alternatives
+
+### `Nullability` on `ITypeSymbol`
+
+The basic principle of this alternative proposal is to add `Nullability Nullability { get; }` to `ITypeSymbol`, and remove the rest of the proposed nullability retrieval APIs in the main section.
+This has a few benefits:
+
+1. Consumers of `ITypeSymbol`s won't have to keep track of where the type symbol came from to retrieve nullability information, they just have to ask the type.
+2. Printing APIs and code generation APIs do not need to have new overloads introduced to allow communication of top-level nullability information.
+3. Consumers of the new API don't have to separately worry about nullability when comparing types, as `string`, `string~`, and `string?` are truly different types.
+
+However, this approach has a few major issues that ultimately break our ability to do this design.
+
+1. This approach will violate principle #1 of our goals for this API, breaking existing analyzers and refactorings extensively.
+A common strategy for these APIs is to call `Compilation.GetTypeByMetadataName`, or one of the similar overloads, and then look for fields, locals, properties, or other similar symbols of the requested type.
+Up to this point, our guidance has been that, for non-generic types, it is perfectly fine to just compare these symbols via `==` if you're looking for exact matches, and that you only need to resort to `OriginalDefinition` for types with generics.
+This will no longer be possible if we were to add `Nullability` to `ITypeSymbol`.
+2. The default `Nullability` of symbols is another pain point.
+When a class is declared, it doesn't have a nullability.
+It's just an `INamedTypeSymbol`.
+It may have constraints that have nullabilities, but the top-level nullability is not a question that makes any sense.
+However, in order to not break analyzers on nullable-unaware code, we will have to make the default nullability `Unknown`, as old C# code will have this as the nullability when `GetTypeInfo` is called on an expression.
+This would further exacerbate issue #1.
+3. This causes massive explosion in type heirarchies.
+If a user got `string?` and called `GetDeclaredMembers()`, the parent of each of these members must point back to `string?`, and not `string` or `string~`.
+
+For these reasons, we have decided not take this approach.
```

Comment:
These reasons don't seem to make a lot of sense to me.  They seem very speculative and are optimizing for a strange view of the world that seems to want to convey that these types are the same, even though the language is heavily pushing the intuition that they're not.  I would have expected this approach if we had gone with, say, an `[Nullable]`-attribute based way of representing this.  But not when we go the path of something like `string?`.  

Note: it's exceptionally strange since we already shipped the `type?` feature already with nullable-value-types.  And We certainly didn't use the `Nullable` enum for that.